agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/utils/SanitizationHelper.java

```diff
@@ -0,0 +1,143 @@
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.agent.internal.exporter.utils;
+
+import com.microsoft.applicationinsights.agent.internal.common.Strings;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SanitizationHelper {
+  public static final int MAX_KEY_LENGTH = 150;
+  public static final int MAX_VALUE_LENGTH = 8192;
+  public static final int MAX_NAME_LENGTH = 1024;
+  public static final int MAX_ID_LENGTH = 512;
+  public static final int MAX_MESSAGE_LENGTH = 32768;
+  public static final int MAX_URL_LENGTH = 2048;
+
+  /** Function to sanitize both key and value in properties. */
+  @SuppressWarnings("ReturnsNullCollection")
+  public static Map<String, String> sanitizeProperties(Map<String, String> properties) {
+    if (properties == null) {
+      return null;
+    }
+    if (!needsSanitizing(properties)) {
+      // this is an optimization to avoid any memory allocation in the normal case
+      return properties;
+    }
+    Map<String, String> sanitized = new HashMap<>();
+    for (Map.Entry<String, String> entry : properties.entrySet()) {
+      String sanitizedKey = sanitizeKey(entry.getKey());
+      String sanitizedValue = sanitizeValue(entry.getValue());
+      if (!Strings.isNullOrEmpty(sanitizedValue)) {
+        String uniqueKey = makeKeyUnique(sanitizedKey, sanitized);
+        sanitized.put(uniqueKey, sanitizedValue);
+      }
+    }
+    return sanitized;
+  }
+
+  private static boolean needsSanitizing(Map<String, String> properties) {
+    for (Map.Entry<String, String> entry : properties.entrySet()) {
+      String key = entry.getKey();
+      String value = entry.getValue();
+      if (Strings.isNullOrEmpty(key)
+          || Strings.isNullOrEmpty(value)
+          || key.length() > MAX_KEY_LENGTH
+          || value.length() > MAX_VALUE_LENGTH) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /** Function to create unique key. */
+  private static String makeKeyUnique(String key, Map<String, ?> map) {
+    if (!map.containsKey(key)) {
+      return key;
+    }
+    String truncatedKey = Strings.truncate(key, MAX_KEY_LENGTH - 3);
+    int candidate = 1;
+    do {
+      key = truncatedKey + candidate;
+      ++candidate;
```

Comment:
what if candidate go to 3+ digit.. that will exceed the limit?

Reply:
increased the truncation length to 6, which will generate 1million unique keys for a given duplicate key. I think that should be good for this usecase :)